src/SSTIOGroup.cpp

```diff
@@ -0,0 +1,824 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of Intel Corporation nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#include "SSTIOGroup.hpp"
+
+#include <algorithm>
+#include <cinttypes>
+#include <iomanip>
+#include <iterator>
+#include <sstream>
+
+#include "Agg.hpp"
+#include "Exception.hpp"
+#include "Helper.hpp"
+#include "MSR.hpp"
+#include "MSRFieldSignal.hpp"
+#include "PlatformTopo.hpp"
+#include "SSTControl.hpp"
+#include "SSTIO.hpp"
+#include "SSTSignal.hpp"
+#include "geopm_debug.hpp"
+#include "geopm_topo.h"
+
+namespace geopm
+{
+    const std::map<std::string, SSTIOGroup::sst_signal_mailbox_raw_s> SSTIOGroup::sst_signal_mbox_info = {
+        { "SST::CONFIG_LEVEL",
+          { SSTIOGroup::SSTMailboxCommand::TURBO_FREQUENCY, 0x00,
+            {{ "LEVEL", { 0x00, 16, 23, 1.0, M_UNITS_NONE,
+                            "SST configuration level", M_SIGNAL_BEHAVIOR_CONSTANT } }}
+          } },
+        { "SST::TURBOFREQ_SUPPORT",
+          { SSTIOGroup::SSTMailboxCommand::TURBO_FREQUENCY, 0x01,
+            {{ "SUPPORTED", { 0x00, 0, 0, 1.0, M_UNITS_NONE,
+                              "SST-TF is supported",
+                              M_SIGNAL_BEHAVIOR_CONSTANT } }}
+          } },
+        { "SST::HIGHPRIORITY_NCORES",
+          { SSTIOGroup::SSTMailboxCommand::TURBO_FREQUENCY, 0x10,
+            {{ "0", { 0x0000, 0, 7, 1.0, M_UNITS_NONE,
+                      "Count of high-priority turbo frequency cores in bucket 0",
+                      M_SIGNAL_BEHAVIOR_CONSTANT } },
```

Comment:
All of these signals are labled as constant.  This implies that they will never change during runtime, is that correct here?  I think these are hardware values that cannot change, but I just wanted to check.

Reply:
That's right. This is just a table that specifies the sst ioctl inputs associated with each signal. The variants are excluded from this table.

Any of the ioctl variants that exist within one type of signal/control have been given different names, like `SST::COREPRIORITY:0 ... SST::COREPRIORITY:3`. So then they're constant with respect to GEOPM signal/control name.